db/blob/db_blob_basic_test.cc

```diff
@@ -367,6 +368,35 @@ TEST_F(DBBlobBasicTest, MultiGetMergeBlobWithPut) {
   ASSERT_EQ(values[2], "v2_0");
 }
 
+TEST_F(DBBlobBasicTest, BlobDBProperties) {
+  Options options = GetDefaultOptions();
+  options.merge_operator = MergeOperators::CreateStringAppendOperator();
+  options.enable_blob_files = true;
+  options.min_blob_size = 0;
+  Reopen(options);
+
+  const size_t numFile = 10;
+  for (size_t i = 0; i < numFile; i++) {
+    ASSERT_OK(Put("key" + std::to_string(i), "0000000000"));
+    ASSERT_OK(Flush());
+  }
+
+  uint64_t intPropertyValue = 0;
+  EXPECT_TRUE(db_->GetIntProperty("rocksdb.num-blob-files", &intPropertyValue));
+  ASSERT_EQ(intPropertyValue, numFile);
+  uint64_t totalBlobFileSize = 0;
+  EXPECT_TRUE(
+      db_->GetIntProperty("rocksdb.total-blob-file-size", &totalBlobFileSize));
+  uint64_t liveBlobFileSize = 0;
+  EXPECT_TRUE(
+      db_->GetIntProperty("rocksdb.live-blob-file-size", &liveBlobFileSize));
+  ASSERT_EQ(totalBlobFileSize, liveBlobFileSize);
+  EXPECT_TRUE(db_->GetIntProperty("rocksdb.estimate-live-data-size",
+                                  &intPropertyValue));
+  std::string propertyValue = "";
+  EXPECT_TRUE(db_->GetProperty("rocksdb.blob-stats", &propertyValue));
```

Comment:
So I think a few more things could be validated here:
1) That the value returned by `total-blob-file-size` / `live-blob-file-size` actually equals the sum of the sizes from the `BlobFileMetaData` objects. You can access the metadata objects in tests like this: https://github.com/facebook/rocksdb/blob/main/db/blob/db_blob_compaction_test.cc#L499-L512
2) That `total-blob-file-size` is computed over all `Version`s and each blob file is counted just once. Of course, we would need multiple `Version`s for this. The simplest way would be to create an iterator (see https://github.com/facebook/rocksdb/wiki/Iterator) using the `NewIterator` API, say, after the fifth `Flush` in the loop. In order to be able to provide a consistent/immutable view of the database at a certain point in time, Iterators keep the `Version` alive. So if you create an iterator after the fifth flush, you would have two versions: version1 with 5 blob files, and version2 with 10 blob files (where version2's 10 files include version1's original 5, enabling us to test the deduplication of blob files).
3) For `estimated-live-data-size`, it would be nice to test a situation where some blob files have garbage in them. The simplest way to generate garbage is to overwrite (using a second `Put`) or delete a key. Here's an example: https://github.com/facebook/rocksdb/blob/main/db/blob/db_blob_compaction_test.cc#L462-L566

To reduce complexity and because some of the above scenarios might contradict/interfere with each other, we might consider creating multiple test cases. Another approach would be to extend existing test cases instead of adding brand new ones; for instance, we might be able to test `estimate-live-data-size` as part of the `DBBlobCompactionTest.TrackGarbage` test case linked above.

Reply:
Thank you so much for the detailed reply.
Previously I tried some hard-coded numbers but I later found the size varies on different platforms (e.g., FB devserver, servers used by CI) so I removed them. I think what you suggest makes more sense :D

Reply:
I have tried both delete and put the same key with a different value but it seems there is no garbage generated (i.e., the garbage bytes is 0). Do you have some ideas of why this happens?
```
  ASSERT_OK(Put("key1", "0000000000"));
  ASSERT_OK(Flush());
  ASSERT_OK(Put("key2", "0000000000"));
  ASSERT_OK(Flush());
  ASSERT_OK(Delete("key2")); // generate some garbage
  ASSERT_OK(Flush());
  // or
  ASSERT_OK(Put("key2", "1111111111")); // generate some garbage
  ASSERT_OK(Flush());
```
Also, when using a delete, # of blob files is also weird because there should 1 file left but the returned value is still 2. Is this expected?

Reply:
Are you compacting the keys afterwards? See e.g. https://github.com/facebook/rocksdb/blob/main/db/blob/db_blob_compaction_test.cc#L491-L496

Technically, it is the compaction process that "generates" any garbage, since it is during compaction that RocksDB reconciles `Put`s and `Delete`s (or multiple `Put`s) for the same key, potentially resulting in unreferenced data in blob files.

Reply:
One more thing: you might want to put the first key-value for both keys in the same file (i.e. don't call `Flush` in between), like in the test case linked above. The reason is that if you turn the entire blob file into garbage, it will become obsolete and will not be a part of the new `Version`.